### PR TITLE
feat(Workspaces): workspace database exploration

### PR DESCRIPTION
**Pull request by cheikhgwane**
Add features for workspace database exploration.

## Changes

Please list / describe the changes in the codebase for the reviewer(s).

- List tables of a workspace database
- Display details on a database table.

## Screenshots / screencast

image

image

### CODE REVIEW

File: src/workspaces/features/DataPreviewDialog/DataPreviewDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+              { name: tableName }
+            )}
+          </p>
+          <DataGrid
+            data={table.sample}
+            fixedLayout={false}
+            className="mt-4"
+            defaultPageSize={5}
+          >
+            {table.columns.map((c: TableColumn, index) => (
```
[the remark this answers is not in the document]
Comment (qgerome) [reply src/workspaces/features/DataPreviewDialog/DataPreviewDialog.tsx:75]: (To be updated with the new version of the schema)

File: src/workspaces/helpers/database.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
+type lang = "R" | "PYTHON";
+
+const trimText = (text: string): string => {
```
Comment (qgerome) [inline src/workspaces/helpers/database.tsx:3]: What if there is a correct indent ?
It's better to write the snippet with the correct indent directly than trying to massage it

File: src/workspaces/helpers/database.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    .map((l) => l.trimStart())
+    .join("\n");
+};
+
+export const getUsageSnippet = (tableName: string, lang: lang) => {
+  let text = "";
+  switch (lang) {
+    case "PYTHON":
+      text = `import os
+              import pandas as pd
```
Comment (qgerome) [inline src/workspaces/helpers/database.tsx:15]: You'll have to remove all indents here to better format the snippet

**Review by qgerome — CHANGES_REQUESTED** [review]

File: src/pages/workspaces/[workspaceSlug]/databases/index.tsx [hunk 4]
```diff
@@ -54,11 +61,12 @@ const WorkspaceDatabasesPage: NextPageWithLayout = (props: Props) => {
       <WorkspaceLayout.PageContent className="space-y-8">
         <DataGrid
           className="overflow-hidden rounded-md bg-white shadow"
-          data={FAKE_WORKSPACE.database.workspaceTables}
+          data={tables.items}
           defaultPageSize={5}
```
[the remark this answers is not in the document]
Comment (cheikhgwane) [reply src/pages/workspaces/[workspaceSlug]/databases/index.tsx:65]: Yes, that works for me

File: src/pages/workspaces/[workspaceSlug]/databases/index.tsx [hunk 5]
```diff
@@ -84,23 +92,17 @@ const WorkspaceDatabasesPage: NextPageWithLayout = (props: Props) => {
...
             className="py-3"
-            accessor="content"
+            accessor="count"
             id="content"
-            label="Content"
+            label="# Rows"
           >
-            {(value) => <span>{`${value} row(s)`}</span>}
+            {(value) => (
+              <span>{`~ ${value} row${value > 0 ? `'s` : ``} `}</span>
```
[the remark this answers is not in the document]
Comment (cheikhgwane) [reply src/pages/workspaces/[workspaceSlug]/databases/index.tsx:100]: I didn't know we can use i18n that! Thanks for the hint.

File: src/workspaces/features/DataPreviewDialog/DataPreviewDialog.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (cheikhgwane) [reply src/workspaces/features/DataPreviewDialog/DataPreviewDialog.tsx:75]: Okay will update it

File: src/pages/workspaces/[workspaceSlug]/databases/[tableId].tsx [hunk 6]
```diff
@@ -54,17 +30,17 @@ const WorkspaceDatabaseTableViewPage: NextPageWithLayout = (props: Props) => {
...
   const { data } = useWorkspaceDatabaseTablePageQuery({
-    variables: { workspaceSlug: router.query.workspaceSlug as string },
+    variables: {
+      workspaceSlug: router.query.workspaceSlug as string,
+      tableName: router.query.tableId as string,
+    },
   });
 
-  if (!data?.workspace) {
+  if (!data?.workspace || !data?.workspace.database) {
```
[the remark this answers is not in the document]
Comment (qgerome) [reply src/pages/workspaces/[workspaceSlug]/databases/[tableId].tsx:39]: Reminder

File: src/pages/workspaces/[workspaceSlug]/databases/[tableId].tsx [hunk 7]
```diff
@@ -129,43 +106,29 @@ const WorkspaceDatabaseTableViewPage: NextPageWithLayout = (props: Props) => {
...
-                    key={index}
-                    className="py-3 font-medium"
-                    name={s.fieldName}
-                    label={s.fieldName}
-                    accessor="sample"
-                  />
-                ))}
-              </DataGrid>
-            }
+            workspaceId={workspace.slug}
```
[the remark this answers is not in the document]
Comment (qgerome) [reply src/pages/workspaces/[workspaceSlug]/databases/[tableId].tsx:130]: Reminder

File: src/pages/workspaces/[workspaceSlug]/databases/[tableId].tsx [hunk 8]
```diff
@@ -182,10 +145,13 @@ export const getServerSideProps = createGetServerSideProps({
...
       query: WorkspaceDatabaseTablePageDocument,
-      variables: { workspaceSlug: ctx.params?.workspaceSlug },
+      variables: {
+        workspaceSlug: ctx.params?.workspaceSlug,
+        tableName: ctx.params?.tableId,
+      },
     });
 
-    if (!data.workspace) {
+    if (!data.workspace || !data.workspace.database.table) {
```
Comment (qgerome) [inline src/pages/workspaces/[workspaceSlug]/databases/[tableId].tsx:154]:
```suggestion
    if (!data.workspace?.database.table) {
```

File: src/pages/workspaces/[workspaceSlug]/databases/index.tsx [hunk 9]
```diff
@@ -24,14 +22,23 @@ type Props = {
...
   const router = useRouter();
-  const { data } = useWorkspaceDatabasesPageQuery({
-    variables: { workspaceSlug: router.query.workspaceSlug as string },
+  const { perPage, page } = props;
+  const workspaceSlug = router.query.workspaceSlug as string;
+  const { data, refetch } = useWorkspaceDatabasesPageQuery({
+    variables: { workspaceSlug: workspaceSlug },
   });
 
+  const onChangePage = ({ page }: { page: number }) => {
```
[the remark this answers is not in the document]
Comment (qgerome) [reply src/pages/workspaces/[workspaceSlug]/databases/index.tsx:31]: Reminder

File: src/workspaces/features/DataPreviewDialog/DataPreviewDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+              sample
+            }
+          }
+        }
+      }
+    `,
+    { variables: { workspaceSlug: workspaceId, tableName: tableName } }
+  );
+
+  if (!data?.workspace) {
```
Comment (qgerome) [inline src/workspaces/features/DataPreviewDialog/DataPreviewDialog.tsx:45]:
```suggestion
  if (!data?.workspace?.database.table) {
```

File: src/workspaces/helpers/database.tsx [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+export const getUsageSnippet = (tableName: string, lang: lang) => {
+  let text = "";
+  switch (lang) {
+    case "PYTHON":
+      text = `import os
+import pandas as pd
+from sqlalchemy import create_engine
+
+engine = create_engine(os.environ["POSTGRESQL_HEXA_EXPLORE_DEMO_URL"])
```
Comment (qgerome) [inline src/workspaces/helpers/database.tsx:11]: I think you could change this by WORKSPACE_DB_URL

**Review by qgerome — CHANGES_REQUESTED** [review]

**Comment by qgerome** [comment]
Don't forget to address this problem (seen in the browser's console)
CleanShot 2023-02-23 at 16 05 55@2x

**Comment by cheikhgwane** [comment]
> Don't forget to address this problem (seen in the browser's console) CleanShot 2023-02-23 at 16 05 55@2x

Hum I cannot reproduce it, weird !

**Comment by cheikhgwane** [comment]
> Don't forget to address this problem (seen in the browser's console) CleanShot 2023-02-23 at 16 05 55@2x

Okay on it

**Review by qgerome — APPROVED** [review]

**Pull request merged**